node/src/components/storage.rs

```diff
@@ -1279,12 +1284,39 @@ impl Storage {
         signature: Box<FinalitySignature>,
     ) -> Result<bool, FatalStorageError> {
         let mut txn = self.env.begin_rw_txn()?;
-        let mut block_signatures = txn
-            .get_value(self.block_metadata_db, signature.block_hash())?
-            .unwrap_or_else(|| BlockSignatures::new(*signature.block_hash(), signature.era_id()));
-        block_signatures.insert_signature(*signature);
-        let outcome = txn.put_value(
-            self.block_metadata_db,
+        let mut block_signatures = self
+            .block_metadata_dbs
+            .get(&mut txn, signature.block_hash())?
+            .unwrap_or_else(|| match &*signature {
+                FinalitySignature::V1(signature) => {
+                    BlockSignaturesV1::new(*signature.block_hash(), signature.era_id()).into()
+                }
+                FinalitySignature::V2(signature) => BlockSignaturesV2::new(
+                    *signature.block_hash(),
+                    signature.block_height(),
+                    signature.era_id(),
+                    signature.chain_name_hash(),
+                )
+                .into(),
+            });
+
+        match (&mut block_signatures, *signature) {
+            (BlockSignatures::V1(ref mut block_signatures), FinalitySignature::V1(signature)) => {
+                block_signatures
+                    .insert_signature(signature.public_key().clone(), *signature.signature());
+            }
+            (BlockSignatures::V2(ref mut block_signatures), FinalitySignature::V2(signature)) => {
+                block_signatures
+                    .insert_signature(signature.public_key().clone(), *signature.signature());
+            }
+            (_, signature) => {
+                return Err(FatalStorageError::VariantMismatch(VariantMismatch(
+                    Box::new(*signature.block_hash()),
+                )));
+            }
```

Comment:
In order to getter visibility of the actual details of the mismatch, I'd suggest the following:
```suggestion
            mismatch => {
                let mismatch = VariantMismatch(Box::new((mismatch.0.clone(), mismatch.1)));
                return Err(FatalStorageError::from(mismatch));
            }
```

Reply:
done